Fix ECS Executor compatibility with Airflow 3.x in try_adopt_task_instances

iamapez opened the pull request:
## Summary
Fixes #58205

The `try_adopt_task_instances` method in the ECS Executor was calling `ti.command_as_list()` which doesn't exist in Airflow 3.x due to Task SDK changes. This caused the scheduler to crash with `AttributeError` when trying to adopt orphaned ECS tasks.

## The Problem
- In Airflow 3.x, the `TaskInstance.command_as_list()` method was removed as part of the Task SDK refactoring
- The ECS Executor's `try_adopt_task_instances` method still calls this non-existent method at line 598
- This causes immediate scheduler failure on startup when using ECS Executor with Airflow 3.x

## The Solution
This PR adds version-aware code that:
- Checks `AIRFLOW_V_3_0_PLUS` to determine the Airflow version
- For Airflow 3.x: Reconstructs the command list manually from task instance attributes
- For Airflow 2.x: Uses the existing `command_as_list()` method

## Testing
- Python syntax validated with `py_compile`
- The fix maintains backward compatibility with Airflow 2.x
- Scheduler starts successfully with this fix applied to Airflow 3.1.1

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas  
- [x] My changes generate no new warnings
- [x] The fix is minimal and focused on the specific issue
- [x] Maintains backward compatibility with Airflow 2.x

iamapez committed:
Fix ECS Executor compatibility with Airflow 3.x in try_adopt_task_instances

The try_adopt_task_instances method was calling ti.command_as_list() which
doesn't exist in Airflow 3.x due to Task SDK changes. This caused scheduler
to crash with AttributeError when trying to adopt orphaned ECS tasks.

This fix:
- Checks AIRFLOW_V_3_0_PLUS to determine Airflow version
- For Airflow 3.x: Reconstructs the command list manually from task instance attributes
- For Airflow 2.x: Uses the existing command_as_list() method

Fixes apache/airflow#58205

iamapez committed:
Fix execution_date renamed to logical_date in Airflow 3.x

In Airflow 3.x, the TaskInstance.execution_date attribute was renamed
to logical_date. This updates the fix to use the correct attribute name.

o-nikolas commented on providers/amazon/src/airflow/providers/amazon/aws/executors/ecs/ecs_executor.py:
```diff
@@ -591,11 +591,31 @@ def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[Task
+                            ti.dag_id,
+                            ti.task_id,
+                            ti.logical_date.isoformat(),
+                        ]
```
This is not how you want to generate a cmd for an Airflow 3.x task. The commands look different in AF3.

See here:  https://github.com/apache/airflow/blob/main/providers/amazon/src/airflow/providers/amazon/aws/executors/ecs/ecs_executor.py#L509-L518

o-nikolas replied:
Could consider making this a helper function now as well, instead of copy/pasting the implementation here.

iamapez committed:
Use proper Task SDK workload serialization for Airflow 3.x adopted tasks

Match the command generation pattern used for newly queued tasks.
Creates ExecuteTask workload, serializes it as JSON, and constructs
the proper SDK command structure.

o-nikolas reviewed (commented):
Is there a unit tests we can add for this?

iamapez committed:
Refactor ECS Executor to consolidate command generation logic

Address code review feedback by consolidating duplicated command generation
into reusable helper methods and adding comprehensive test coverage.

Changes:
- Add _serialize_workload_to_command() static method for workload serialization
- Add _build_task_command() method for version-aware command generation
- Refactor execute_async() to use helper method (removed 9 lines of duplication)
- Refactor try_adopt_task_instances() to use helper method (removed 18 lines)
- Add comprehensive unit tests for both Airflow 2.x and 3.x compatibility
- Update test_try_adopt_task_instances and remove skip marker for Airflow 3.x
- Add Tradepost Markets Inc to INTHEWILD.md

Test coverage:
- test_serialize_workload_to_command: validates workload-to-command conversion
- test_build_task_command_airflow3: validates Airflow 3.x command generation
- test_build_task_command_airflow2: validates Airflow 2.x command generation
- test_try_adopt_task_instances: updated for Airflow 3.x support

iamapez committed:
Merge branch 'main' into fix-ecs-executor-airflow3-compatibility

iamapez committed:
Merge branch 'main' into fix-ecs-executor-airflow3-compatibility

iamapez committed:
Merge branch 'main' into fix-ecs-executor-airflow3-compatibility

iamapez committed:
Merge branch 'main' into fix-ecs-executor-airflow3-compatibility

iamapez committed:
Merge branch 'main' into fix-ecs-executor-airflow3-compatibility

o-nikolas commented on providers/amazon/src/airflow/providers/amazon/aws/exceptions.py:
```diff
@@ -17,7 +17,7 @@
 from __future__ import annotations
 
-from airflow.providers.common.compat.sdk import AirflowException
+from airflow.exceptions import AirflowException
```
Removing this compat doesn't look right, why did you make this change?

iamapez committed:
Merge branch 'main' into fix-ecs-executor-airflow3-compatibility
